libs/ml/src/dataloaders/tensor_dataloader.cpp

```diff
@@ -49,6 +49,27 @@ typename TensorDataLoader<TensorType>::ReturnType TensorDataLoader<TensorType>::
   return ReturnType(ret_labels, ret_data);
 }
 
+template <typename TensorType>
+OperationsCount TensorDataLoader<TensorType>::ChargeGetNext()
+{
+  OperationsCount cost = 1;
+  for (auto it : one_sample_label_shape_)
+  {
+    cost *= it;
+  }
```

Comment:
it would be good to add comments explaining exactly which operations in GetNext incur this cost.
It's difficult for me to tell why this might be correct, so some explanation is absolutely necessary.
Presumably in this case this loop should have the comment for this line?:
"  TensorType ret_labels = labels_.View(*this->current_cursor_).Copy(one_sample_label_shape_); "

Also, you don't need to write for loops like this for vectors; the math library has a Product function for doing this which means less repeated code and probably a slightly better optimisation.